src/geometry/euclidean.lean

```diff
@@ -920,4 +920,328 @@ calc dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) * dist (r1 • v +ᵥ p1) (r2 
+  (∀ p1 p2 ∈ ps, dist p1 p = dist p2 p) ↔ ∃ r, ∀ p1 ∈ ps, dist p1 p = r :=
+begin
+  split,
+  { intro h,
+    by_cases hn : ps.nonempty,
+    { cases hn with q hq,
+      exact ⟨dist q p, λ p1 hp1, h p1 q hp1 hq⟩ },
+    { rw set.not_nonempty_iff_eq_empty at hn,
+      use 0,
+      intros p1 hp1,
+      rw hn at hp1,
+      exact false.elim hp1 } },
+  { rintros ⟨r, hr⟩ p1 p2 hp1 hp2,
+    rw [hr p1 hp1, hr p2 hp2] }
+end
+
+include V
+
+/-- There exists `r` such that `p` has distance `r` from all the
+points of a set of points in `s` if and only if there exists (possibly
+different) `r` such that its `orthogonal_projection` has that distance
+from all the points in that set. -/
+lemma exists_dist_eq_iff_exists_dist_orthogonal_projection_eq {s : affine_subspace ℝ P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {ps : set P}
+    (hps : ps ⊆ s) (p : P) :
+  (∃ r, ∀ p1 ∈ ps, dist p1 p = r) ↔
+    ∃ r, ∀ p1 ∈ ps, dist p1 (orthogonal_projection hn hc p) = r :=
+begin
+  have h := dist_set_eq_iff_dist_orthogonal_projection_eq hn hc hps p,
+  simp_rw dist_set_eq_iff_exists_dist_eq at h,
+  exact h
+end
+
+/-- The induction step for the existence and uniqueness of the
+circumcenter.  Given a nonempty set of points in a nonempty affine
+subspace whose direction is complete, such that there is a unique
+(circumcenter, circumradius) pair for those points in that subspace,
+and a point `p` not in that subspace, there is a unique (circumcenter,
+circumradius) pair for the set with `p` added, in the span of the
+subspace with `p` added. -/
+lemma exists_unique_dist_eq_of_insert {s : affine_subspace ℝ P} (hn : (s : set P).nonempty)
+    (hc : is_complete (s.direction : set V)) {ps : set P} (hnps : ps.nonempty) {p : P}
+    (hps : ps ⊆ s) (hp : p ∉ s)
+    (hu : ∃! cccr : (P × ℝ), cccr.fst ∈ s ∧ ∀ p1 ∈ ps, dist p1 cccr.fst = cccr.snd) :
+  ∃! cccr₂ : (P × ℝ), cccr₂.fst ∈ affine_span ℝ (insert p (s : set P)) ∧
+    ∀ p1 ∈ insert p ps, dist p1 cccr₂.fst = cccr₂.snd :=
+begin
+  rcases hu with ⟨⟨cc, cr⟩, ⟨hcc, hcr⟩, hcccru⟩,
+  simp only [prod.fst, prod.snd] at hcc hcr hcccru,
+  let x := dist cc (orthogonal_projection hn hc p),
+  let y := dist p (orthogonal_projection hn hc p),
+  have hy0 : y ≠ 0 := dist_orthogonal_projection_ne_zero_of_not_mem hn hc hp,
+  let ycc₂ := (x * x + y * y - cr * cr) / (2 * y),
+  let cc₂ := (ycc₂ / y) • (p -ᵥ orthogonal_projection hn hc p : V) +ᵥ cc,
+  let cr₂ := real.sqrt (cr * cr + ycc₂ * ycc₂),
+  use (cc₂, cr₂),
+  simp only [prod.fst, prod.snd],
+  have hpo :
+    p = (1 : ℝ) • (p -ᵥ orthogonal_projection hn hc p : V) +ᵥ orthogonal_projection hn hc p,
+  { simp },
+  split,
+  { split,
+    { refine vadd_mem_of_mem_direction _ (mem_affine_span ℝ (set.mem_insert_of_mem _ hcc)),
+      rw direction_affine_span,
+      exact submodule.smul_mem _ _
+        (vsub_mem_vector_span ℝ (set.mem_insert _ _)
+                                (set.mem_insert_of_mem _ (orthogonal_projection_mem hn hc _))) },
```

Comment:
Random remark: may of these lines/names are getting very long. What do you think of renaming (for example) `orthogonal_projection` to `ortho_proj`?
Of course this can be done in another PR.

Reply:
`orthogonal_projection` isn't even one of the longer names; it's more a matter of e.g. `dist_square_eq_dist_orthogonal_projection_square_add_dist_orthogonal_projection_square`. I'm not sure what conventions are for when a simple application of mathlib naming conventions produces such a long name.

Reply:
But if you rename `orthogonal_projection` to `ortho_proj` you save `2 * 11 = 22` characters on that long name. That's quite substantial, I think.

Reply:
It seems reasonable to rename as long as the version for inner product spaces is renamed at the same time. It wouldn't help with some related long names not mentioning `orthogonal_projection`, such as those for various versions of Pythagoras, however.